iib/workers/tasks/opm_operations.py

```diff
@@ -918,3 +922,254 @@ def verify_operator_exists(
+        binary_image,
+        '--from-index',
+        from_index,
+        '--operators',
+        ','.join(operators),
+    ]
+
+    log.info(
+        'Generating the database file from an existing database %s and excluding'
+        ' the following operator(s): %s',
+        from_index,
+        ', '.join(operators),
+    )
+
+    with set_registry_token(overwrite_from_index_token, from_index, append=True):
+        run_cmd(cmd, {'cwd': base_dir}, exc_msg='Failed to remove operators from the index image')
+
+
+def deprecate_bundles(
+    bundles: List[str],
+    base_dir: str,
+    binary_image: str,
+    from_index: str,
+    overwrite_target_index_token: Optional[str] = None,
+    container_tool: Optional[str] = None,
+) -> None:
+    """
+    Deprecate the specified bundles from the index image.
+
+    Only Dockerfile is created, no build is performed.
+
+    :param list bundles: pull specifications of bundles to deprecate.
+    :param str base_dir: base directory where operation files will be located.
+    :param str binary_image: binary image to be used by the new index image.
+    :param str from_index: index image, from which the bundles will be deprecated.
+    :param str overwrite_target_index_token: the token used for overwriting the input
+        ``from_index`` image. This is required to use ``overwrite_target_index``.
+        The format of the token must be in the format "user:password".
+    :param str container_tool: the container tool to be used to operate on the index image
+    """
+    from iib.workers.tasks.utils import run_cmd, set_registry_token
+
+    cmd = [
+        opm_path,
+        'index',
+        'deprecatetruncate',
+        '--generate',
+        '--binary-image',
+        binary_image,
+        '--from-index',
+        from_index,
+        '--bundles',
+        ','.join(bundles),
+        '--allow-package-removal',
+    ]
+    if container_tool:
+        cmd.append('--container-tool')
+        cmd.append(container_tool)
+    with set_registry_token(overwrite_target_index_token, from_index):
+        run_cmd(cmd, {'cwd': base_dir}, exc_msg='Failed to deprecate the bundles')
+
+
+def _find_index_version(index_image: str) -> str:
+    """
+    Find the OCP version of the given index image.
+
+    :param str index_image: index_image pull spec
+    :return: version of index image
+    :rtype: str
+
+    """
+    from iib.workers.tasks.utils import get_image_label
+
+    index_image_version = index_image.split(":")[1]
+    if index_image_version.startswith("v4."):
+        return index_image_version
+    else:
+        index_version = get_image_label(index_image, 'com.redhat.index.delivery.version')
+        return index_version
+
+
+def set_opm_verison(from_index: Optional[str] = None):
+    """
+    Set the opm version to be used for the entire IIB operation.
+
+    opm version is based on from_index/source_from_index.
+
+    :param str from_index: from_index_image for the request
+    """
+    global opm_path
```

Comment:
Do we have to set a global variable? Can we set a default in the base `Config` object that gets overwritten if a value is set in iib-playbooks?

Reply:
I don't think in this case we need to update anything at config level, but can you give an example how that could help?

Reply:
The approach I suggested was to set `opm_path='opm'` in the `Config` class as default. Then, you change that variable on line 1146 if the user has specified `opm_paths_config`. That should eliminate the need of having a global variable.

Reply:
Its still unclear how that can benefit instead of using a global variable, I meant the example of same in the earlier comment.
Also, config variables generally remain same for the entire service and for all request types, so changing a config variable in every request seemed a bit odd

Reply:
Apologies my previous comment wasn't clear enough.

Right, I am not suggesting to change the config variable during every request. 

In opm_operations.py, `opm_path='opm'` is being initialized in 3 places. Let's say if we have to change the way the opm command is used in the future, like `/usr/bin/opm`, we would have to replace it in 3 places. Having a config variable makes it easier to change, can avoid potential bugs in the future and make the code easier to understand. My suggestion is to use `opm_path=worker_config.opm_path` and then modify `opm_path`

Secondly, using global variable generally makes the code difficult to understand and maintain especially when it's being updated inside the functions. Maybe we can think about creating a class with `__init__`, `get_opm_path` and `set_opm_path` functions? 

@chandwanitulsi @lipoja @xDaile thoughts on the class?

Reply:
I am in favour of using the class object here, as it would clear some uncertainties related to the usage of global variables.

Reply:
I agree, we should have one variable with default "opm" path. And it might be part of worker config.
And yes, there are better ways then using global variable.  So the ideal to have it as class attribute sounds good.

Reply:
So I read a lot online regarding the use of global variables and class attr, it looks like there are strong opinions outside for both approaches. 
To go with majority I am exploring the class attribute way. I also read that get() and set() methods are not ideal way as well as we can effectively use class.attr or obj.attr directly without needing to have getters and setters (using property rather than getter and setter sounded like a preferred approach), therefore I am avoiding the complete get/set way but using the set(), also with this approach, we will have to access the attr everywhere by class.attr

For the other approach Michal mentioned about putting opm_*() functions inside the class as well, we can definitely do that, but then in that case, some functions like opm_index_add, opm_index_rm, and deprecate_bundles will become instance methods, but there will still be several other functions in the file, which does not need to be instance methods and IMO it would be better to keep those outside the class, but that could make the file structure a bit confusing

Reply:
Sorry, my terminology was not good. What I meant with "class object" was the creation of the OPM class instance with incoming requests, where, for example, the OPM version could be set during the __init__() method. I mentioned moving the methods under this class, as it would allow us to use the instance opm version automatically.